File: SLL.cpp
#include <iostream>

class Node
{
public:
    int value;
    Node *next;
    Node(int value) : value(value), next(nullptr) {}
};

class LinkedList
{
private:
    Node *head;
    Node *tail;
    int length;

public:
    LinkedList(int value) : head(new Node(value)), tail(head), length(1) {}

    void printList() const
    {
        Node *current = head;

        while (current != nullptr)
        {
            std::cout << current->value << std::endl;
            current = current->next;
        }
    }

    void append(int value)
    {
        Node *newNode = new Node(value);
        if (isEmpty())
        {
            head = tail = newNode;
        }
        else
        {
            tail->next = newNode;
            tail = newNode;
        }
        length++;
    }

    void prepend(int value)
    {
        Node *newNode = new Node(value);
        if (length == 0)
        {
            head = tail = newNode;
        }
        else
        {
            newNode->next = head;
            head = newNode;
        }
        length++;
    }

    void deletelast()
    {
        if (length == 0)
            return;

        Node *current = head;
        if (length == 1)
        {
            delete head;
            head = tail = nullptr;
        }
        else
        {
            while (current->next != tail)
            {
                current = current->next;
            }
            delete tail;
            tail = current;
            tail->next = nullptr;
        }

        length--;
    }

    bool isEmpty() const
    {
        return length == 0;
    }
};

int main()
{
    LinkedList *SLL = new LinkedList(5);
    SLL->printList();
    SLL->append(11);
    SLL->printList();
}

File: data structures/c#/linked-list/SLL.cs
using System;
class Node
{
    public int Value;
    public Node Next;

    public Node(int value)
    {
        Value = value;
        Next = null;
    }
}


class LinkedList
{
    private Node Head;
    private Node Tail;
    private int Length;

    public LinkedList()
    {
        Head = null;
        Tail = null;
        Length = 0;
    }

    public void PrintValues()
    {
        Node current = Head;

        while (current != null)
        {
            Console.WriteLine(current.Value);
            current = current.Next;
        }
    }

    public bool IsEmpty()
    {
        return Length == 0;
    }

    //The Append method adds a new node to the end of the list, updating Tail and the list's length.
    public void Append(int value)
    {
        Node newNode = new Node(value);

        if (Head == null)
        {
            Head = newNode;
            Tail = newNode;
        }
        else
        {
            Tail.Next = newNode;
            Tail = newNode;
        }

        Length++;
    }
}


class Program
{
    public static void Main()
    {
        LinkedList list = new LinkedList();

        list.Append(10);
        list.Append(20);
        list.Append(30);

        list.PrintValues();

    }

}